cpp/src/plasma/protocol.cc

```diff
@@ -230,35 +230,63 @@ Status ReadReleaseReply(uint8_t* data, size_t size, ObjectID* object_id) {
   return plasma_error_status(message->error());
 }
 
-// Delete messages.
+// Delete objects messages.
 
-Status SendDeleteRequest(int sock, ObjectID object_id) {
+Status SendDeleteRequest(int sock, const std::vector<ObjectID>& object_ids) {
   flatbuffers::FlatBufferBuilder fbb;
-  auto message = CreatePlasmaDeleteRequest(fbb, fbb.CreateString(object_id.binary()));
+  auto message =
+      CreatePlasmaDeleteRequest(fbb, static_cast<int>(object_ids.size()),
+                                to_flatbuffer(&fbb, &object_ids[0], object_ids.size()));
   return PlasmaSend(sock, MessageType::PlasmaDeleteRequest, &fbb, message);
 }
 
-Status ReadDeleteRequest(uint8_t* data, size_t size, ObjectID* object_id) {
+Status ReadDeleteRequest(uint8_t* data, size_t size, std::vector<ObjectID>& object_ids) {
   DCHECK(data);
-  auto message = flatbuffers::GetRoot<PlasmaReleaseReply>(data);
+  auto message = flatbuffers::GetRoot<PlasmaDeleteRequest>(data);
   DCHECK(verify_flatbuffer(message, data, size));
-  *object_id = ObjectID::from_binary(message->object_id()->str());
+  int num_objects = message->count();
+  object_ids.clear();
+  object_ids.reserve(num_objects);
+  for (int i = 0; i < num_objects; ++i) {
+    object_ids.push_back(ObjectID::from_binary(message->object_ids()->Get(i)->str()));
+  }
   return Status::OK();
 }
 
-Status SendDeleteReply(int sock, ObjectID object_id, PlasmaError error) {
+Status SendDeleteReply(int sock, const std::vector<ObjectID>& object_ids,
+                       const std::vector<PlasmaError>& errors) {
+  DCHECK(object_ids.size() == errors.size());
   flatbuffers::FlatBufferBuilder fbb;
-  auto message =
-      CreatePlasmaDeleteReply(fbb, fbb.CreateString(object_id.binary()), error);
+  auto message = CreatePlasmaDeleteReply(
+      fbb, static_cast<int>(object_ids.size()),
+      to_flatbuffer(&fbb, &object_ids[0], object_ids.size()),
+      fbb.CreateVector(reinterpret_cast<const int*>(&errors[0]), object_ids.size()));
   return PlasmaSend(sock, MessageType::PlasmaDeleteReply, &fbb, message);
 }
 
-Status ReadDeleteReply(uint8_t* data, size_t size, ObjectID* object_id) {
+Status ReadDeleteReply(uint8_t* data, size_t size, std::vector<ObjectID>& object_ids,
+                       std::vector<PlasmaError>& errors) {
   DCHECK(data);
   auto message = flatbuffers::GetRoot<PlasmaDeleteReply>(data);
   DCHECK(verify_flatbuffer(message, data, size));
-  *object_id = ObjectID::from_binary(message->object_id()->str());
-  return plasma_error_status(message->error());
+  int num_objects = message->count();
+  object_ids.clear();
+  object_ids.reserve(num_objects);
+  errors.clear();
+  errors.reserve(num_objects);
+  for (int i = 0; i < num_objects; ++i) {
+    object_ids.push_back(ObjectID::from_binary(message->object_ids()->Get(i)->str()));
+  }
+  PlasmaError return_error = PlasmaError::ObjectNonexistent;
```

Comment:
Again this code leaves me pretty confused. Why return an error only if _all_ objects are nonexistent? Thinking about it more (similar to the above comment), I would again say if a non-existing object is deleted, it shouldn't be an error.

This would make delete a hint to the plasma store that an object may be deleted, which I think is the most intuitive semantics. We can discuss more in a JIRA.